src/envs/tools.py

```diff
@@ -0,0 +1,643 @@
+            num_items = num_items_lst[i % len(num_items_lst)]
+            num_contraptions = num_contraptions_lst[i %
+                                                    len(num_contraptions_lst)]
+            data = {}
+            # Initialize robot
+            data[self._robot] = np.array([1.0])  # fingers start off open
+            contraptions: List[Object] = []
+            # Initialize contraptions
+            for j in range(num_contraptions):
+                contraption = Object(f"contraption{j}", self._contraption_type)
+                while True:
+                    pose_lx = rng.uniform(
+                        self.table_lx, self.table_ux - self.contraption_size)
+                    pose_ly = rng.uniform(
+                        self.table_ly, self.table_uy - self.contraption_size)
+                    pose_ux = pose_lx + self.contraption_size
+                    pose_uy = pose_ly + self.contraption_size
+                    # Make sure no other contraption intersects with this one
+                    if all(data[other][2] < pose_lx or \
+                           data[other][0] > pose_ux or \
+                           data[other][3] < pose_ly or \
+                           data[other][1] > pose_uy for other in contraptions):
+                        break
+                contraptions.append(contraption)
+                data[contraption] = np.array(
+                    [pose_lx, pose_ly, pose_ux, pose_uy], dtype=np.float32)
+            # Initialize items (screws, nails, bolts) and set goal
+            # We enforce that there can only be at most one screw, to make
+            # the problems generally easier to solve
+            items: List[Object] = []
+            screw_cnt, nail_cnt, bolt_cnt = 0, 0, 0
+            screw_created = False
+            goal = set()
+            for _ in range(num_items):
+                while True:
+                    pose_x = rng.uniform(self.table_lx, self.table_ux)
+                    pose_y = rng.uniform(self.table_ly, self.table_uy)
+                    # If collision with any contraption, try again
+                    if any(data[c][0] < pose_x < data[c][2] and \
+                           data[c][1] < pose_y < data[c][3]
+                           for c in contraptions):
+                        continue
+                    # If collision with any item, try again
+                    if any(abs(data[i][0] - pose_x) < self.close_thresh and \
+                           abs(data[i][1] - pose_y) < self.close_thresh
+                           for i in items):
+                        continue
+                    # Otherwise, we found a valid pose
+                    break
+                on_table = 1.0  # always start off on table
+                is_fastened = 0.0  # always start off not fastened
+                is_held = 0.0  # always start off not held
+                choices = ["screw", "nail", "bolt"]
+                if screw_created:
```

Comment:
can't we just check `screw_cnt`?

Reply:
yes thanks, the code went through a lot of revisions...